Sources/graphql-swift/codegen/jsonUtils.swift

```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+public protocol InitializableFromJSON {
+    init(_ json: [String: Any])
+}
+
+public extension Dictionary where Key == String {
+    func val(_ key: String) -> [String: Any]? {
+        return self[key] as? [String:Any]
+    }
+    
+    func arr(_ key: String) -> [[String: Any]]? {
+        return self[key] as? [[String:Any]]
+    }
+    
+    func get<T>(_ key: String) -> T? {
+        return self[key] as? T
+    }
+}
+
+public extension Array where Element == Dictionary<String, Any> {
+    func produce<T>() -> [T]? where T: InitializableFromJSON {
+        return self.map { T($0) }
```

Comment:
Tip: This can be written as:
```suggestion
        return map(T(_:))
```

Reply:
Whoa. Not needing `self.` is obvious in retrospect. But why don't I need to pass the value I'm mapping over (ie $0)?

Ohhhh ... wait is `T(:)` a *reference to* the default initializer for T? So I'm calling map, and passing it whatever T's initializer is that, takes a single argument?

Reply:
You’re correct! Although please see the updated code I forgot the `.init` part, so it’s `T.init(_:)` 😅  This also works with other functions of the form `A -> B`. There’s some fun composition libraries around that adds things like concat, curry etc. E.g. https://github.com/pointfreeco/swift-overture

Reply:
Closures, free functions and methods are all the same in Swift and can be interchanged freely. A common pattern where this is used is [type erasure](https://www.donnywals.com/understanding-type-erasure-in-swift/).